src/coordinator/functions/temporal/base.go

```diff
@@ -0,0 +1,348 @@
+
+	processRequests := make([]processRequest, 0, len(leftBlks))
+	// If we have all blocks for the left range in the cache, then process the current block
+	if !emptyLeftBlocks {
+		processRequests = append(processRequests, processRequest{blk: b, deps: leftBlks, bounds: bounds})
+	}
+
+	leftBlks = append(leftBlks, b)
+
+	// Process right side of the range
+	rightBlks, emptyRightBlocks := c.processRight(b, bounds, maxBlocks, rightRangeStart)
+
+	for i := 0; i < len(rightBlks); i++ {
+		lStart := maxBlocks - i
+		if lStart > len(leftBlks) {
+			continue
+		}
+
+		deps := leftBlks[len(leftBlks)-lStart:]
+		deps = append(deps, rightBlks[:i]...)
+		processRequests = append(processRequests, processRequest{blk: rightBlks[i], deps: deps, bounds: bounds.Next(i + 1)})
+
+	}
+
+	// If either the left range or right range wasn't fully processed then cache the current block
+	if emptyLeftBlocks || emptyRightBlocks {
+		if err := c.cache.Add(bounds.Start, b); err != nil {
+			return err
+		}
+	}
+
+	return c.processCompletedBlocks(processRequests, queryStartBounds, queryEndBounds, maxBlocks)
+}
+
+// processLeft processes the current block. For the current block, figure out whether we have enough previous blocks which can help process it
+func (c *baseNode) processLeft(b block.Block, bounds block.Bounds, maxBlocks int, leftRangeStart time.Time) ([]block.Block, bool) {
+	leftRangeTimes := make([]time.Time, 0, maxBlocks)
+	for t := leftRangeStart; t.Before(bounds.Start); t = t.Add(bounds.Duration) {
+		leftRangeTimes = append(leftRangeTimes, t)
+	}
+
+	leftBlks := c.cache.MultiGet(leftRangeTimes)
+	lastNil := lastEmpty(leftBlks)
+	if lastNil >= 0 {
+		return leftBlks[lastNil:], true
+	}
+
+	return leftBlks, false
+}
+
+// processRight processes blocks after current block. This is done by fetching all contiguous right blocks until the right range
+func (c *baseNode) processRight(b block.Block, bounds block.Bounds, maxBlocks int, rightRangeStart time.Time) ([]block.Block, bool) {
+	rightRangeTimes := make([]time.Time, 0, maxBlocks)
+	for t := bounds.End(); !t.After(rightRangeStart); t = t.Add(bounds.Duration) {
+		rightRangeTimes = append(rightRangeTimes, t)
+	}
+
+	rightBlks := c.cache.MultiGet(rightRangeTimes)
+	firstNil := firstEmpty(rightBlks)
+	return rightBlks[:firstNil], firstNil != len(rightBlks)
+}
+
+// processCompletedBlocks processes all blocks for which are dependant blocks are present
+func (c *baseNode) processCompletedBlocks(processRequests []processRequest, queryStartBounds, queryEndBounds block.Bounds, maxBlocks int) error {
+	processedKeys := make([]time.Time, len(processRequests))
+	for i, req := range processRequests {
+		if err := c.processSingleRequest(req); err != nil {
+			return err
+		}
+
+		processedKeys[i] = req.bounds.Start
+	}
+
+	// Mark all blocks as processed
+	c.cache.MarkProcessed(processedKeys)
+	// Sweep to free blocks from cache with no dependencies
+	c.sweep(c.cache.Processed(), queryStartBounds, queryEndBounds, maxBlocks)
```

Comment:
does sweep need to take in `cache.Processed()` if it's a method on `c` already?

Reply:
keeping sweep simple ?